File: data structure/queue/queue.c
#include <stdio.h>
#define QUEUE_SIZE 5

// typedef struct
// {

//     int head, tail;
//     int data[QUEUE_SIZE + 1];

// } Queue;

// void Enqueue(Queue *q, int item)
// {

//     if ((q->tail + 1) % (QUEUE_SIZE + 1) == q->head)
//     {
//         printf("Queue is Full!!\n");
//         return;
//     }

//     q->data[q->tail] = item;
//     q->tail = (q->tail + 1) % (QUEUE_SIZE + 1);
// }

// int Dequeue(Queue *q)
// {
//     if (q->head == q->tail)
//     {
//         printf("Queue is Empty!!\n");
//         return -1;
//     }
//     else
//     {

//         int item = q->data[q->head];
//         q->data[q->head] = -1;
//         q->head = (q->head + 1) % (QUEUE_SIZE + 1);
//         return item;
//     }
// }

typedef struct
{
    int data[QUEUE_SIZE + 1];
    int head, tail;
} Queue;

void Enqueue(Queue *q, int item)
{

    if ((q->tail + 1) % (QUEUE_SIZE + 1) == q->head)
    {
        printf("Queue is full!\n");
        return;
    }

    q->data[q->tail] = item;
    q->tail = (q->tail + 1) % (QUEUE_SIZE + 1);
}

int Dequeue(Queue *q)
{

    if (q->head == q->tail)
    {
        printf("Queue is empty!!\n");
        return -1;
    }

    int item = q->data[q->head];
    q->head = (q->head + 1) % (QUEUE_SIZE + 1);
    return item;
}

int main()
{
    int i = 0;
    Queue new_Queue;
    new_Queue.head = 0;
    new_Queue.tail = 0;

    Enqueue(&new_Queue, 10);
    Enqueue(&new_Queue, 20);
    Enqueue(&new_Queue, 30);
    Enqueue(&new_Queue, 40);
    Enqueue(&new_Queue, 50);
    Enqueue(&new_Queue, 60);

    Dequeue(&new_Queue);
    Enqueue(&new_Queue, 50);
    Enqueue(&new_Queue, 50);

    printf("head:%d  tail:%d\n", new_Queue.head, new_Queue.tail);

    return 0;
}

File: data structure/queue/queue.js
class Queue {
  constructor(size) {
    this.QueueSize = size;
    this.head = 0;
    this.tail = 0;
    this.data = [this.QueueSize + 1];
  }

  Enqueue(item) {
    if ((this.tail + 1) % (this.QueueSize + 1) == this.head) {
      console.log("Queue is Full!!");
      return;
    }

    this.data[this.tail] = item;
    this.tail = (this.tail + 1) % (this.QueueSize + 1);
  }

  Dequeue() {
    if (this.head == this.tail) {
      console.log("Queue is Empty!!");
      return 0;
    }

    const item = this.data[this.head];
    this.head = (this.head + 1) % (this.QueueSize + 1);
    return item;
  }
}

const q1 = new Queue(5);
q1.Enqueue(10);
q1.Enqueue(20);
q1.Enqueue(30);
q1.Enqueue(40);
q1.Enqueue(50);
q1.Dequeue();
q1.Dequeue();
q1.Dequeue();
q1.Dequeue();
q1.Dequeue();

q1.Enqueue(50);
q1.Enqueue(50);
q1.Enqueue(50);
q1.Enqueue(50);
q1.Enqueue(50);

console.log(`head: ${q1.head} , tail: ${q1.tail}`);
